src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,267 @@ PhaseStatus Compiler::optUnrollLoops()
+    // When inside a loop, branches are quicker than selects.
+    // Detect via the block weight as that will be high when inside a loop.
+    if (block->getBBWeight(this) > BB_UNITY_WEIGHT)
+    {
+        return false;
+    }
+
+    // Does the block end by branching via a JTRUE after a compare?
+    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    {
+        return false;
+    }
+
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* last = block->lastStmt()->GetRootNode();
+    noway_assert(last->gtOper == GT_JTRUE);
+    if (!last->AsOp()->gtOp1->OperIsCmpCompare() || (last->gtFlags & GTF_SIDE_EFFECT) != 0 ||
+        (last->AsOp()->gtOp1->gtFlags & GTF_COLON_COND) != 0)
+    {
+        return false;
+    }
+
+    BasicBlock* middle_block = block;
+    GenTree*    asg_node     = nullptr;
+    Statement*  asg_stmt     = nullptr;
+    bool        found_select = false;
+
+    // Check the the block is followed by a block or chain of blocks that only contain NOPs and
+    // a single ASG statement. The destination of the final block must point to the same as the
+    // true path of the JTRUE block.
+    bool found_middle = false;
+    while (!found_middle)
+    {
+        middle_block = middle_block->bbNext;
+        noway_assert(middle_block != nullptr);
+
+        if (middle_block->bbNext == block->bbJumpDest)
+        {
+            // This is our final middle block.
+            found_middle = true;
+        }
+
+        // Make sure there is only one block which jumps to the middle block,
+        // and the middle block is not the start of a TRY block or an exception handler.
+        noway_assert(!fgCheapPredsValid);
+        if (middle_block->NumSucc() != 1 || middle_block->bbJumpKind != BBJ_NONE ||
+            middle_block->bbPreds->flNext != nullptr || middle_block->bbCatchTyp != BBCT_NONE ||
+            ((middle_block->bbFlags & (BBF_TRY_BEG | BBF_DONT_REMOVE)) != 0))
+        {
+            return false;
+        }
+
+        // Can all the nodes within the middle block be made to conditionally execute?
+        for (Statement* const stmt : middle_block->Statements())
+        {
+            GenTree* tree = stmt->GetRootNode();
+            switch (tree->gtOper)
+            {
+                case GT_ASG:
+                {
+                    GenTree* op1 = tree->AsOp()->gtOp1;
+                    GenTree* op2 = tree->AsOp()->gtOp2;
+
+                    // Only one per assignment per block can be conditionally executed.
+                    // Ensure the destination of the assign is a local variable with integer type,
+                    // and the nodes of the assign won't cause any additional side effects.
+                    if (asg_node != nullptr || op1->gtOper != GT_LCL_VAR || !varTypeIsIntegralOrI(op1->TypeGet()) ||
+                        (op1->gtFlags & GTF_SIDE_EFFECT) != 0 || (op2->gtFlags & GTF_SIDE_EFFECT) != 0)
```

Comment:
```suggestion
                    if (asg_node != nullptr || op1->OperIs(GT_LCL_VAR) || !varTypeIsIntegralOrI(op1) ||
                        (op1->gtFlags & GTF_SIDE_EFFECT) != 0 || (op2->gtFlags & GTF_SIDE_EFFECT) != 0)
```

Also, I'm not sure that this handles volatile loads correctly, please do a few tests with those.